lib/src/metta/runner/pkg_mgmt/catalog.rs

```diff
@@ -89,60 +93,200 @@ use xxhash_rust::xxh3::xxh3_64;
+                        }
+                    },
+                    None => {
+                        ret_desc = Some(desc);
+                        highest_version = Some(ver)
+                    }
+                }
+            },
+            None => {
+                if highest_version.is_none() {
+                    if let Some(ret_desc) = ret_desc {
+                        log::warn!("Multiple un-versioned {} modules in catalog; impossible to select newest", ret_desc.name());
+                    }
+                    ret_desc = Some(desc)
+                }
+            }
+        }
+    }
+    ret_desc
 }
 
 /// The object responsible for locating and selecting dependency modules for each [MettaMod]
 ///
 /// This structure is conceptually analogous to the a `Cargo.toml` file for a given module.
-#[derive(Clone, Debug, Default)]
-//TODO: Use serde to deserialize a PkgInfo from an expression atom
+#[derive(Clone, Debug, Default, Deserialize)]
 pub struct PkgInfo {
 
     /// The public name of the module.  Should be composed of alpha-numeric characters with '-' and '_'
     /// characters allowed.  Must not contain any other punctuation. 
-    pub name: String,
+    pub name: Option<String>,
 
-    //TODO: version field, to indicate the version of this module
+    // The version of this module
+    #[serde(default)]
+    pub version: Option<semver::Version>,
 
     /// If `strict == true` then a dependency must be declared in the `PkgInfo`, otherwise a permissive
     /// version requirement will be assumed for any modules that are not explicitly declared
+    #[serde(default)]
     pub strict: bool,
 
     /// Entries mapping module names to requirements for each dependency sub-module
     ///
     /// A Duplicate entry for a given sub-module in the deps list is an error.
+    #[serde(default)]
     pub deps: HashMap<String, DepEntry>,
 }
 
 /// A single entry in a [PkgInfo]'s dependencies, specifying the properties of a module that will satisfy a dependency
-#[derive(Clone, Debug, Default)]
+#[derive(Clone, Debug, Default, Deserialize)]
 pub struct DepEntry {
     /// Indicates that the dependency module should be loaded from a specific FS path
     ///
     /// If the fs_path is specified, the other pkg_info attributes will be ignored.
-    //QUESTION: We need a MeTTa "style guide" for these field names, since they are effective going
-    // to be part of the API, because a PkgInfo will be deserialized from atoms
-    pub fs_path: Option<PathBuf>
+    #[serde(default)]
+    pub fs_path: Option<PathBuf>,
 
-    //TODO: field for fetching from a specific git repo
-    //TODO: field to indicate acceptable version range for dependency
+    #[serde(flatten)]
+    git_location: ModuleGitLocation,
+
+    /// An acceptable version of version bounds to satisfy the dependency.  None means any version
+    /// acceptable
+    #[serde(default)]
+    pub version_req: Option<semver::VersionReq>
```

Comment:
If there is a justification for serialization format to be different from internal representation (as you provided above) it is possible to make it different. On the other hand I prefer internal representation which reflects the nature of the data. For example to me having both `file_path` and `git URL` set at the same time doesn't make sense but current structure allows it. Also having both `file_path` and `version` doesn't make sense to me as the only version can be kept at some file path. Finally the version or tag or branch could be a part of Git URL (which is not standard I believe but used frequently) thus I am also not sure about having version and Git URL set at the same time.

Reply:
> having both file_path and git URL set at the same time doesn't make sense but current structure allows it.

I agree that it's not ideal, but I think it's the lesser-evil for the reasons outlined above.  ie. forcing people to use a tagged nested structure.

> Also having both file_path and version doesn't make sense to me as the only version can be kept at some file path.

If the module at at the specified path / git location has an incompatible version, then the load should fail.  I think a version requirement should exist irrespective of a location.

> Finally the version or tag or branch could be a part of Git URL

I haven't tested what happens when the git URL encodes a branch.  I have tested setting the `git_branch` field https://github.com/luketpeterson/hyperon-experimental/blob/0fb5cc29a96ff6d90f63a17c48cbb1752fa36df6/lib/src/metta/runner/pkg_mgmt/git_catalog.rs#L27C9-L27C19 and that does work.

I should probably add keys for tag & revision

Cargo generates a warning if the URL specifies any embedded tag, branch, or revision.  https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#choice-of-commit

Authentication is another can of worms.  It's getting to be less and less important with people putting their credentials into their profiles, but neither Cargo nor this PR have a feature to password authentication, and I don't think it would be possible to add one.

Reply:
> I haven't tested what happens when the git URL encodes a branch.

I believe it doesn't work out of the box as it is not standard. Some systems support it but they probably have to parse URL manually and use API.

> Authentication is another can of worms

I don't think it is needed actually. If someone wants to use private repo he is able to use the local copy of the repo.